### DISCUSSION TITLE
0079 XLS-79d:  Transaction Fee RPC

### DISCUSSION THREAD

**Discussion by tequdev** [XRPLF/XRPL-Standards#225] [XLS Proposals]
## 1. Abstract

Fees required for a transaction to be accepted into the network depends on the type of transaction and network conditions.

Currently some client libraries such as xrpl.js calculate Fees based on the type of transaction and its fields, but this proposal extends Fee RPC to provide a specification for calculating the Fee required by a transaction without a client library.

Furthermore, in a multi-chain ecosystem with the XRPL protocol, it would be ideal to have multiple networks available in one client library, and even if the network's own transactions require special fees (e.g, for Memo fields or other dynamic fields), it should be possible to obtain the fee.

## 2. Overview

Add a `tx_blob` and `tx_json` field to the already existing `fee` RPC request and modify it to return the Fee information required for transactions in this field to be accepted into the network.

This change does not require an Amendment because it is only an RPC change that does not affect consensus or transaction processing.

### 3.1. New Request Fields

Add a new optional field to the existing Fee RPC request:

| Field Name | JSON Type | Description |
| ---------- | --------- | ----------- |
| `tx_blob`  | `string`  | (Optional) The transaction to get fee, in [binary format](https://xrpl.org/docs/references/protocol/binary-format/). |
| `tx_json`  | `object`  | (Optional) The transaction to get fee, in JSON format. |

If either `tx_blob` or `tx_json` is included, the RPC behaves as per the existing `fee` RPC specification.

Requirements for the `tx_blob` or `tx_json`:
- Contains either `tx_blob` or `tx_json` or neither
- Valid transaction format

### 3.2. New Response Fields

No change in the response field.

However, if the request field contains `tx_blob` or `tx_json`, fields in `result.drops` will be values to which the following are taken into account

- base fee
- signerCount
- Special transaction types (e.g.,`AccountDelete`, `AMMCreate`, `LedgerStateFix`)
- `Fulfillment` field of `EscrowCreate`

### 3.3. Error Cases

The RPC should return error messages for invalid `tx_blob` or `tx_json` inputs if those fields are included.

### 4.1. Request

```json
{
  "id": 1,
  "command": "fee",
  "tx_blob": "1200152400000000684000000000000000730081142269E8F72D5112F6992818E2B8EC2309EFB9D3C08314046D18FE322220FB5538108533015D485EA2D12E"
}
```

```json
{
  "id": 1,
  "command": "fee",
  "tx_json": {
    "TransactionType": "AccountDelete",
    "Account": "rh3xiHDPSKvvyFVVYHufgUb7useHNukPr8",
    "Destination": "rQQQrUdN1cLdNmxH4dHfKgmX5P4kf3ZrM"
  }
}
```

### 4.2. Response

```json
{
  "id": 1,
  "result": {
    "current_ledger_size": "114",
    "current_queue_size": "0",
    "drops": {
      "base_fee": "200000",
      "median_fee": "100000000",
      "minimum_fee": "200000",
      "open_ledger_fee": "200000"
    },
    "expected_ledger_size": "168",
    "ledger_current_index": 8386800,
    "levels": {
      "median_level": "128000",
      "minimum_level": "256",
      "open_ledger_level": "256",
      "reference_level": "256"
    },
    "max_queue_size": "3360"
  },
  "type": "response"
}
```

## 5. Backwards Compatibility

This proposal is backwards compatible as it only extends the existing `fee` RPC without modifying its current behavior when `tx_blob` is not provided.

**Comment by dangell7** [comment]
`open_ledger_fee` should also be updated for the txn type.

**Comment by mvadari** [comment]
IMO this should follow the same parameter style as `simulate`, which autofills `Sequence`, `Fee`, and `SigningPubKey`, and also accepts a `tx_json` in place of the `tx_blob`.

**Reply by tequdev** [reply]
It'd be simpler to support only tx_blob within this standard proposal, as it'd need to be changed for each additional autofill field in the future.

**Reply by mvadari** [reply]
> It'd be simpler to support only tx_blob within this standard proposal

I agree that it'd be simpler, but I don't see why that's a concern. IMO if we're making the change, we should do it completely, rather than halfway.

> it'd need to be changed for each additional autofill field in the future

Autofill fields aren't added very often, so I also don't see how that's a concern. Worst case scenario, you just have to autofill it yourself before making the call (until it's patched). Also, adding `tx_json` doesn't change anything there.

**Reply by mvadari** [reply]
Another note - there's a lot of code that can be reused from `simulate`/adjusted to be helper functions for all of the transaction-ubmission-related RPCs.

**Comment by tequdev** [comment]
Updated
- Added `tx_json` field to request parameter

[a comment by an automated account is not shown]
